File: src/logan/pickpocket/user/PickpocketUser.java
package logan.pickpocket.user;

import com.sk89q.worldedit.bukkit.BukkitAdapter;
import com.sk89q.worldedit.util.Location;
import com.sk89q.worldguard.LocalPlayer;
import com.sk89q.worldguard.WorldGuard;
import com.sk89q.worldguard.bukkit.WorldGuardPlugin;
import com.sk89q.worldguard.protection.regions.RegionContainer;
import com.sk89q.worldguard.protection.regions.RegionQuery;
import logan.config.MessageConfiguration;
import logan.pickpocket.main.PickpocketPlugin;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.entity.Player;

import java.util.UUID;


/**
 * Created by Tre on 12/14/2015.
 */
public class PickpocketUser {

    private UUID uuid;
    private PickpocketUser victim;
    private PickpocketUser predator;
    private PickpocketUser recentPredator;
    private boolean playingMinigame;
    private boolean rummaging;
    private RummageInventory openRummageInventory;
    private boolean participating;

    private ProfileConfiguration profileConfiguration;
    private MinigameModule minigameModule;

    public PickpocketUser(UUID uuid) {
        this.uuid = uuid;

        profileConfiguration = new ProfileConfiguration(PickpocketPlugin.getInstance().getDataFolder() + "/players/", uuid.toString() + ".yml");
        profileConfiguration.createSections();

        participating = profileConfiguration.getParticipatingSectionValue();

        minigameModule = new MinigameModule(this);
    }

    public void performPickpocket(PickpocketUser victim) {

        minigameModule = new MinigameModule(this);

        boolean isAllowedPickpocketing = true;
        if (PickpocketPlugin.isWorldGuardPresent()) {
            LocalPlayer localPlayer = WorldGuardPlugin.inst().wrapPlayer(getPlayer());
            RegionContainer container = WorldGuard.getInstance().getPlatform().getRegionContainer();
            RegionQuery query = container.createQuery();
            Location worldEditLocation = BukkitAdapter.adapt(getPlayer().getLocation());
            isAllowedPickpocketing = query.testState(worldEditLocation, localPlayer, PickpocketPlugin.PICKPOCKET_FLAG);
        }

        if (!isAllowedPickpocketing) {
            getPlayer().sendMessage(PickpocketPlugin.getMessageConfiguration().getMessage(MessageConfiguration.PICKPOCKET_REGION_DISALLOW_KEY));
            return;
        }
        if (!PickpocketPlugin.getCooldowns().containsKey(uuid)) {
            openRummageInventory = new RummageInventory(victim);
            openRummageInventory.show(this);
            setRummaging(true);
            setVictim(victim);
            victim.setPredator(this);
        } else {
            getPlayer().sendMessage(PickpocketPlugin.getMessageConfiguration().getMessage(MessageConfiguration.COOLDOWN_NOTICE_KEY, PickpocketPlugin.getCooldowns().get(uuid).toString()));
        }
    }

    public RummageInventory getOpenRummageInventory() {
        return openRummageInventory;
    }

    public boolean isPlayingMinigame() {
        return playingMinigame;
    }

    public boolean isRummaging() {
        return rummaging;
    }

    public void setRummaging(boolean value) {
        rummaging = value;
    }

    public void setPlayingMinigame(boolean value) {
        playingMinigame = value;
    }

    public void setVictim(PickpocketUser victim) {
        this.victim = victim;
    }

    public PickpocketUser getVictim() {
        return victim;
    }

    public void setPredator(PickpocketUser predator) {
        this.predator = predator;
        if (predator != null) {
            recentPredator = predator;
        }
    }

    public PickpocketUser getPredator() {
        return predator;
    }

    public PickpocketUser getRecentPredator() {
        return recentPredator;
    }

    public boolean isPredator() {
        return victim != null;
    }

    public boolean isVictim() {
        return predator != null;
    }

    public void setParticipating(boolean participating) {
        this.participating = participating;
        profileConfiguration.setParticipatingSection(participating);
    }

    public boolean isParticipating() {
        return participating;
    }

    public void setPlayer(UUID uuid) {
        this.uuid = uuid;
    }

    public Player getPlayer() {
        return Bukkit.getPlayer(uuid);
    }

    public UUID getUuid() {
        return uuid;
    }

    public ProfileConfiguration getProfileConfiguration() {
        return profileConfiguration;
    }

    public MinigameModule getMinigameModule() {
        return minigameModule;
    }

    public void sendMessage(String message, Object... args) {
        getPlayer().sendMessage(ChatColor.translateAlternateColorCodes('&', String.format(message, args)));
    }

    public void playRummageSound() {
        getPlayer().playSound(getPlayer().getLocation(), PickpocketPlugin.getAPIWrapper().getSoundBlockSnowStep(), 1.0f, 0.5f);
    }

    @Override
    public String toString() {
        return "player: " + (uuid != null ? getPlayer().getName() : "null") + "\nvictim: " + (victim != null ? victim.getPlayer().getName() : "null")
                + "\npredator: " + (predator != null ? predator.getPlayer().getName() : "null") + "\nrecent predator: " + (recentPredator != null ? recentPredator.getPlayer().getName() : "null"
                + "\nmini-game: " + playingMinigame + "\nrummaging: " + rummaging + "\nparticipating: " + participating);
    }
}

File: src/logan/pickpocket/user/PickpocketUser.kt
package logan.pickpocket.user

import logan.pickpocket.config.MessageConfiguration
import logan.pickpocket.main.PickpocketPlugin
import logan.pickpocket.main.ThiefProfile
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.Sound
import org.bukkit.entity.Player
import java.util.*

class PickpocketUser(val uuid: UUID) {
    val bukkitPlayer
        get() = Bukkit.getPlayer(uuid)
    var victim: PickpocketUser? = null
    var predator: PickpocketUser? = null
        set(value) {
            field = value
            if (lastPredator == null) lastPredator = field
        }
    var lastPredator: PickpocketUser? = null
    val isPredator
        get() = victim != null
    val isVictim
        get() = predator != null
    var isPlayingMinigame = false
    var isRummaging = false
    var openRummageInventory: RummageInventory? = null
    var isAdmin
        get() = playerConfiguration.isAdmin
        set(value) { playerConfiguration.isAdmin = value }
    var isBypassing
        get() = playerConfiguration.isBypassing
        set(value) { playerConfiguration.isBypassing = value }
    var isExempt
        get() = playerConfiguration.isExempt
        set(value) { playerConfiguration.isExempt = value }
    var isParticipating
        get() = playerConfiguration.isParticipating
        set(value) { playerConfiguration.isParticipating = value }
    var currentMinigame: Minigame? = null
    val playerConfiguration =
        PlayerConfiguration("${PickpocketPlugin.instance.dataFolder}/players/", "$uuid.yml")
    var thiefProfile
        get() = findThiefProfile()
        set(value) { playerConfiguration.thiefProfile = value?.name ?: "default" }
    var steals
        get() = playerConfiguration.stealCount
        set(value) { playerConfiguration.stealCount = value }

    fun doPickpocket(victim: PickpocketUser) {
        when {
            !WorldGuardUtil.isPickpocketingAllowed(bukkitPlayer!!) -> {
                bukkitPlayer!!.sendMessage(MessageConfiguration.pickpocketRegionDisallowMessage)
            }
            isCoolingDown() -> bukkitPlayer!!.sendMessage(
                MessageConfiguration.getCooldownNoticeMessage(PickpocketPlugin.getCooldowns()[bukkitPlayer].toString())
            )
            else -> {
                this.victim = victim
                victim.predator = this
                openRummageInventory = RummageInventory(victim)
                openRummageInventory?.show(this)
                isRummaging = true
            }
        }
    }

    private fun findThiefProfile(): ThiefProfile? {
        return PickpocketPlugin.profileConfiguration.loadThiefProfiles().find { bukkitPlayer!!.hasPermission("pickpocket.profile.thief.${it.name}") }
            ?: PickpocketPlugin.profileConfiguration.loadThiefProfiles().find { thiefProfile?.equals(it) ?: return null }
    }

    fun assignThiefProfile(name: String): Boolean {
        PickpocketPlugin.profileConfiguration.loadThiefProfile(name)?.run {
            thiefProfile = this
            return true
        } ?: return false
    }

    fun giveCooldown() {
        val thiefProfile = findThiefProfile() ?: return
        if (!playerConfiguration.isBypassing) PickpocketPlugin.addCooldown(
            bukkitPlayer!!,
            thiefProfile.cooldown
        ) else Unit
    }

    private fun isCoolingDown() = PickpocketPlugin.getCooldowns().containsKey(bukkitPlayer)

    fun sendMessage(message: String, vararg args: Any) {
        bukkitPlayer?.sendMessage(ChatColor.translateAlternateColorCodes('&', String.format(message, args)))
    }

    fun playRummageSound() {
        with(bukkitPlayer) {
            this?.playSound(
                location,
                Sound.BLOCK_SNOW_STEP,
                1.0f,
                0.5f
            )
        }
    }

    companion object {
        fun get(player: Player): PickpocketUser {
            for (user in PickpocketPlugin.users) {
                if (user.uuid == player.uniqueId) {
                    return user
                }
            }
            val user = PickpocketUser(player.uniqueId)
            PickpocketPlugin.addProfile(user)
            return user
        }
    }
}